File: nodejs/test/e2e/llm_inference.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll, type LlmInferenceRequest } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

/**
 * Drain the request body and reply with a single buffered response. The
 * unified callback supports both buffered and streaming uniformly — for
 * non-streaming responses, the consumer writes the whole body once and
 * calls `end`.
 */
async function respondBuffered(
    req: LlmInferenceRequest,
    init: { status: number; headers?: Record<string, string[]> },
    body: string,
): Promise<void> {
    for await (const _chunk of req.requestBody) {
        // discard — the runtime always sends at least one chunk (with end:true).
    }
    await req.responseBody.start(init);
    if (body.length > 0) {
        await req.responseBody.write(body);
    }
    await req.responseBody.end();
}

async function handleNonStreaming(req: LlmInferenceRequest): Promise<void> {
    const url = req.url.toLowerCase();
    if (url.endsWith("/models")) {
        return respondBuffered(
            req,
            { status: 200, headers: { "content-type": ["application/json"] } },
            JSON.stringify({
                data: [
                    {
                        id: "claude-sonnet-4.5",
                        name: "Claude Sonnet 4.5",
                        object: "model",
                        vendor: "Anthropic",
                        version: "1",
                        preview: false,
                        model_picker_enabled: true,
                        capabilities: {
                            type: "chat",
                            family: "claude-sonnet-4.5",
                            tokenizer: "o200k_base",
                            limits: { max_context_window_tokens: 200000, max_output_tokens: 8192 },
                            supports: { streaming: true, tool_calls: true, parallel_tool_calls: true, vision: true },
                        },
                    },
                ],
            }),
        );
    }
    if (url.includes("/models/session")) {
        return respondBuffered(req, { status: 200, headers: {} }, "{}");
    }
    if (url.includes("/policy")) {
        return respondBuffered(req, { status: 200, headers: {} }, JSON.stringify({ state: "enabled" }));
    }
    return respondBuffered(
        req,
        { status: 200, headers: { "content-type": ["application/json"] } },
        "{}",
    );
}

describe("LLM inference callback", async () => {
    const received: LlmInferenceRequest[] = [];

    const { copilotClient: client } = await createSdkTestContext({
        copilotClientOptions: {
            llmInference: {
                createLlmInferenceProvider: () => ({
                    async onLlmRequest(req): Promise<void> {
                        received.push(req);
                        await handleNonStreaming(req);
                    },
                }),
            },
        },
    });

    it("registers the provider on connect without erroring", async () => {
        await client.start();
        expect(client).toBeDefined();
    });

    it(
        "invokes the callback for non-streaming model-layer requests and threads sessionId through",
        async () => {
            const baselineLength = received.length;
            const session = await client.createSession({ onPermissionRequest: approveAll });
            try {
                // Drive a turn so model-layer traffic (catalog,
                // session-intent, inference) flows through the callback.
                // We swallow errors here — the buffered handler returns
                // empty JSON for inference, which is not a valid model
                // response; the agent will surface a transport error.
                // What we care about is that the runtime *attempted* to
                // call the callback for the model-layer endpoints.
                try {
                    await session.sendAndWait({ prompt: "Say OK." });
                } catch {
                    // expected — see comment above
                }
            } finally {
                await session.disconnect();
            }

            expect(received.length).toBeGreaterThan(baselineLength);
            const newRequests = received.slice(baselineLength);
            for (const r of newRequests) {
                expect(r.url).toMatch(/^https?:\/\//);
                expect(typeof r.method).toBe("string");
            }

            const catalog = newRequests.find((r) => r.url.toLowerCase().endsWith("/models"));
            expect(catalog, "expected to intercept the /models catalog request").toBeDefined();

            const inSession = newRequests.find((r) => typeof r.sessionId === "string");
            if (inSession) {
                expect(inSession.sessionId).toMatch(/[a-zA-Z0-9-]+/);
            }
        },
        90_000,
    );
});

File: go/internal/e2e/llm_inference_e2e_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package e2e

import (
	"strings"
	"sync"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// llmRecordingHandler answers every model-layer request with the synthetic
// non-inference fallback (catalog / session / policy, and empty JSON for the
// inference call itself). It records what it intercepted.
type llmRecordingHandler struct {
	mu       sync.Mutex
	received []*copilot.LlmInferenceRequest
}

func (h *llmRecordingHandler) OnLlmRequest(req *copilot.LlmInferenceRequest) error {
	h.mu.Lock()
	h.received = append(h.received, req)
	h.mu.Unlock()
	return llmHandleNonInferenceModelTraffic(req, nil)
}

func (h *llmRecordingHandler) snapshot() []*copilot.LlmInferenceRequest {
	h.mu.Lock()
	defer h.mu.Unlock()
	return append([]*copilot.LlmInferenceRequest(nil), h.received...)
}

func TestLlmInferenceCallback(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	handler := &llmRecordingHandler{}
	client := newLlmClient(ctx, handler)
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}

	// The buffered fallback returns empty JSON for the inference call, which is
	// not a valid model response, so the turn fails; swallow that. What we
	// assert is that the runtime attempted the callback.
	_, _ = session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."})
	_ = session.Disconnect()

	received := handler.snapshot()
	if len(received) == 0 {
		t.Fatal("Expected the runtime to invoke the inference callback")
	}

	var sawCatalog bool
	for _, r := range received {
		if !strings.HasPrefix(r.URL, "http://") && !strings.HasPrefix(r.URL, "https://") {
			t.Fatalf("Expected an absolute URL, got %q", r.URL)
		}
		if strings.HasSuffix(strings.ToLower(r.URL), "/models") {
			sawCatalog = true
		}
		if r.SessionID != "" && len(r.SessionID) == 0 {
			t.Fatal("session id should be non-empty when present")
		}
	}
	if !sawCatalog {
		t.Fatal("Expected to intercept the /models catalog request")
	}
}
